<a href="https://colab.research.google.com/github/P4RASTOO/Challenge_12/blob/main/credit_risk_resampling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Credit Risk Classification

Credit risk poses a classification problem that’s inherently imbalanced. This is because healthy loans easily outnumber risky loans. In this Challenge, you’ll use various techniques to train and evaluate models with imbalanced classes. You’ll use a dataset of historical lending activity from a peer-to-peer lending services company to build a model that can identify the creditworthiness of borrowers.

## Instructions:

This challenge consists of the following subsections:

* Split the Data into Training and Testing Sets

* Create a Logistic Regression Model with the Original Data

* Predict a Logistic Regression Model with Resampled Training Data

### Split the Data into Training and Testing Sets

Open the starter code notebook and then use it to complete the following steps.

1. Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

2. Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

    > **Note** A value of `0` in the “loan_status” column means that the loan is healthy. A value of `1` means that the loan has a high risk of defaulting.  

3. Check the balance of the labels variable (`y`) by using the `value_counts` function.

4. Split the data into training and testing datasets by using `train_test_split`.

### Create a Logistic Regression Model with the Original Data

Employ your knowledge of logistic regression to complete the following steps:

1. Fit a logistic regression model by using the training data (`X_train` and `y_train`).

2. Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

3. Evaluate the model’s performance by doing the following:

    * Calculate the accuracy score of the model.

    * Generate a confusion matrix.

    * Print the classification report.

4. Answer the following question: How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

### Predict a Logistic Regression Model with Resampled Training Data

Did you notice the small number of high-risk loan labels? Perhaps, a model that uses resampled data will perform better. You’ll thus resample the training data and then reevaluate the model. Specifically, you’ll use `RandomOverSampler`.

To do so, complete the following steps:

1. Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points.

2. Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

3. Evaluate the model’s performance by doing the following:

    * Calculate the accuracy score of the model.

    * Generate a confusion matrix.

    * Print the classification report.
    
4. Answer the following question: How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

### Write a Credit Risk Analysis Report

For this section, you’ll write a brief report that includes a summary and an analysis of the performance of both machine learning models that you used in this challenge. You should write this report as the `README.md` file included in your GitHub repository.

Structure your report by using the report template that `Starter_Code.zip` includes, and make sure that it contains the following:

1. An overview of the analysis: Explain the purpose of this analysis.


2. The results: Using bulleted lists, describe the balanced accuracy scores and the precision and recall scores of both machine learning models.

3. A summary: Summarize the results from the machine learning models. Compare the two versions of the dataset predictions. Include your recommendation for the model to use, if any, on the original vs. the resampled data. If you don’t recommend either model, justify your reasoning.

In [1]:
# Install the required libraries
%%capture
!pip install scikit-learn

In [2]:
# Import the modules
import numpy as np
import pandas as pd
from google.colab import drive
from sklearn.metrics import classification_report
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from imblearn.metrics import classification_report_imbalanced

import warnings
warnings.filterwarnings('ignore')

%matplotlib inline

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [3]:
from google.colab import drive
drive.mount('/content/drive')

df_lending = pd.read_csv('/content/drive/MyDrive/lending_data.csv')

# Review the DataFrame
df_lending

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


loan_size  interest_rate  borrower_income  debt_to_income  \
0        10700.0          7.672            52800        0.431818   
1         8400.0          6.692            43600        0.311927   
2         9000.0          6.963            46100        0.349241   
3        10700.0          7.664            52700        0.430740   
4        10800.0          7.698            53000        0.433962   
...          ...            ...              ...             ...   
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  loan_status  
0                    5                 1       22800            0  
1                    3                 0       13600            0  
2                    3                 0       16100            0  
3                    5                 1       22700            0  
4                    5                 1       23000            0  
...                ...               ...         ...          ...  
77531               12                 2       56600            1  
77532               11                 2       50900            1  
77533               11                 2       50300            1  
77534               10                 2       45300            1  
77535                9                 2       42300            1  

[77536 rows x 8 columns]

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [4]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = df_lending['loan_status']

# Separate the X variable, the features
X = df_lending.drop(columns=['loan_status'])

In [5]:
# Review the y variable Series
print(y)

0        0
1        0
2        0
3        0
4        0
        ..
77531    1
77532    1
77533    1
77534    1
77535    1
Name: loan_status, Length: 77536, dtype: int64


In [6]:
# Review the X variable DataFrame
X

loan_size  interest_rate  borrower_income  debt_to_income  \
0        10700.0          7.672            52800        0.431818   
1         8400.0          6.692            43600        0.311927   
2         9000.0          6.963            46100        0.349241   
3        10700.0          7.664            52700        0.430740   
4        10800.0          7.698            53000        0.433962   
...          ...            ...              ...             ...   
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  
0                    5                 1       22800  
1                    3                 0       13600  
2                    3                 0       16100  
3                    5                 1       22700  
4                    5                 1       23000  
...                ...               ...         ...  
77531               12                 2       56600  
77532               11                 2       50900  
77533               11                 2       50300  
77534               10                 2       45300  
77535                9                 2       42300  

[77536 rows x 7 columns]

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

> Indented block



In [7]:
# Check the balance of our target values
label_counts = y.value_counts()
print(label_counts)

0    75036
1     2500
Name: loan_status, dtype: int64


### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [8]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)


---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [9]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
model = LogisticRegression(random_state=1)

# Fit the model using training data
model.fit(X_train, y_train)

LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [10]:
# Make a prediction using the testing data
predicted_y = model.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [11]:
# Print the balanced_accuracy score of the model
balanced_acc_score = balanced_accuracy_score(y_test, predicted_y)
balanced_acc_score

0.9521352751368186

In [12]:
# Generate a confusion matrix for the model
confusion_matrix(y_test, predicted_y)

array([[14926,    75],
       [   46,   461]])

In [13]:
# Print the classification report for the model
target_names = ["Repaid", "Defaulted"]
print('Classification Report:')
print(classification_report(y_test, predicted_y, target_names=target_names))

Classification Report:
              precision    recall  f1-score   support

      Repaid       1.00      1.00      1.00     15001
   Defaulted       0.86      0.91      0.88       507

    accuracy                           0.99     15508
   macro avg       0.93      0.95      0.94     15508
weighted avg       0.99      0.99      0.99     15508



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The precision for both labels is quite high, indicating that the model has a low false positive rate. For label 0, the precision is 1.00, meaning that when the model predicts a loan as healthy (0), it is almost always correct. For label 1, the precision is 0.86, indicating that when the model predicts a loan as high-risk (1), it is correct around 86% of the time.

The recall is also high for both labels, indicating that the model can correctly identify a significant proportion of positive instances (high-risk loans) and negative instances (healthy loans) in the dataset. A recall of 1.00 for label 0 means that the model correctly identifies all healthy loans in the dataset, and a recall of 0.91 for label 1 indicates that it correctly identifies around 91% of high-risk loans.

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points.

In [14]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
oversampler = RandomOverSampler(random_state=1)


# Fit the original training data to the random_oversampler model
X_train_resampled, y_train_resampled = oversampler.fit_resample(X_train, y_train)


In [15]:
# Count the distinct values of the resampled labels data
counts_resampled = pd.Series(y_train_resampled).value_counts()
counts_resampled

0    60035
1    60035
Name: loan_status, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [16]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
model_resampled = LogisticRegression(random_state=1)

# Fit the model using the resampled training data
model_resampled.fit(X_train_resampled, y_train_resampled)

# Make a prediction using the testing data
predicted_y_resampled = model_resampled.predict(X_test)


### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [17]:
# Print the balanced_accuracy score of the model
print(balanced_accuracy_score(y_test, predicted_y_resampled))

0.9941749445500477


In [18]:
# Generate a confusion matrix for the model
confusion_matrix(y_test, predicted_y_resampled)


array([[14915,    86],
       [    3,   504]])

In [19]:
# Print the classification report for the model
print('Classification Report Resampled:')
print(classification_report(y_test, predicted_y_resampled))

Classification Report Resampled:
              precision    recall  f1-score   support

           0       1.00      0.99      1.00     15001
           1       0.85      0.99      0.92       507

    accuracy                           0.99     15508
   macro avg       0.93      0.99      0.96     15508
weighted avg       1.00      0.99      0.99     15508



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The precision of 1.00 for label 0 means that when the model predicts a loan as healthy (0), it is almost always correct. The recall of 0.99 suggests that the model correctly identifies around 99% of healthy loans in the dataset. The F1-score of 1.00 indicates that the model achieves a perfect balance between precision and recall for label 0.

The precision of 0.85 for label 1 indicates that when the model predicts a loan as high-risk (1), it is correct around 85% of the time. The recall of 0.99 suggests that the model correctly identifies around 99% of high-risk loans in the dataset. The F1-score of 0.92 suggests a good balance between precision and recall for label 1.

The logistic regression model fit with oversampled data appears to predict both the 0 (healthy loan) and 1 (high-risk loan) labels quite well.